File: snuqs/circuit/creg.py
from __future__ import annotations


class Creg:
    def __init__(self, name: str, dim: int):
        if dim == 0:
            raise ValueError("Creg of dim 0 is not allowed")

        self.name = name
        self.dim = dim
        self.cbits = [Cbit(self, i) for i in range(dim)]

    def __getitem__(self, key: int):
        return self.cbits[key]

    def __repr__(self):
        return f"{self.name}"


class Cbit:
    def __init__(self, creg: Creg, index: int):
        self.creg = creg
        self.index = index
        self.value = 0
        self.dim = 1

    def __repr__(self):
        return f"{self.creg.name}[{self.index}]"

File: src/circuit/creg.cc
#include "circuit/creg.h"

#include <iostream>
#include <stdexcept>

namespace snuqs {

Creg::Creg() {}
Creg::Creg(std::string name, size_t num_bits)
    : name_(name), num_bits_(num_bits), bitset_{0} {
  if (num_bits == 0) {
    throw std::invalid_argument("Creg of dimension 0 is not allowed");
  }

  if (num_bits > Creg::MAX_BITS) {
    throw std::invalid_argument("Creg dimension is too large");
  }

  for (auto i = 0; i < num_bits; ++i) {
    cbits_.push_back(new Cbit(*this, i, bitset_[i]));
  }
}

Creg::~Creg() {
  std::cout << "Creg Destructor\n";
  for (auto p : cbits_)
    delete p;
}

std::string Creg::name() const { return name_; }
size_t Creg::value() const { return bitset_.to_ulong(); }

Cbit &Creg::__getitem__(size_t key) { return *cbits_[key]; }
void Creg::__setitem__(size_t key, bool val) { bitset_[key] = val; }

Cbit::Cbit(const Creg &creg, size_t index, Creg::bitset::reference bit)
    : creg_(creg), index_(index), bit_(bit) {}

Cbit::~Cbit() {}

} // namespace snuqs
